test/dynamo/test_ctx_manager.py

```diff
@@ -1644,6 +1716,1099 @@ def f(x):
+                yield x + 1
+            finally:
+                pass
+
+        def fn(x):
+            ctx = whoo(x)
+            try:
+                y = ctx.__enter__()
+            finally:
+                ctx.__exit__(None, None, None)
+            return y
+
+        x = torch.tensor([1.0])
+        expected = fn(x)
+
+        eager = EagerAndRecordGraphs()
+        out = torch.compile(backend=eager, fullgraph=False)(fn)(x)
+        self.assertEqual(expected, out)
+        self.assertEqual(len(eager.graphs), 0)
+
+    @unittest.expectedFailure
+    def test_graph_break_after___enter__(self):
+        @contextlib.contextmanager
+        def whoo(x):
+            try:
+                yield x + 1
+            finally:
+                pass
+
+        def fn(x):
+            ctx = whoo(x)
+            try:
+                y = ctx.__enter__()
+                torch._dynamo.graph_break()
+            finally:
+                ctx.__exit__(None, None, None)
+            return y
+
+        x = torch.tensor([1.0])
+        expected = fn(x)
+
+        eager = EagerAndRecordGraphs()
+        out = torch.compile(backend=eager, fullgraph=False)(fn)(x)
+        self.assertEqual(expected, out)
+        self.assertEqual(len(eager.graphs), 0)
+
+    @unittest.expectedFailure
+    def test_graph_break_before_and_after___enter__(self):
+        @contextlib.contextmanager
+        def whoo(x):
+            try:
+                yield x + 1
+            finally:
+                pass
+
+        def fn(x):
+            ctx = whoo(x)
+            try:
+                torch._dynamo.graph_break()
+                y = ctx.__enter__()
+                torch._dynamo.graph_break()
+            finally:
+                ctx.__exit__(None, None, None)
+            return y
+
+        x = torch.tensor([1.0])
+        expected = fn(x)
+
+        eager = EagerAndRecordGraphs()
+        out = torch.compile(backend=eager, fullgraph=False)(fn)(x)
+        self.assertEqual(expected, out)
+        self.assertEqual(len(eager.graphs), 0)
+
+    @unittest.expectedFailure
+    def test_graph_break_before___enter___and_disable___exit__(self):
+        @contextlib.contextmanager
+        def whoo(x):
+            try:
+                yield x + 1
+            finally:
+                pass
+
+        def fn(x):
+            ctx = whoo(x)
+            try:
+                torch._dynamo.graph_break()
+                y = ctx.__enter__()
+            finally:
+
+                @torch._dynamo.disable
+                def g():
+                    ctx.__exit__(None, None, None)
+
+                g()
+            return y
+
+        x = torch.tensor([1.0])
+        expected = fn(x)
+
+        eager = EagerAndRecordGraphs()
+        out = torch.compile(backend=eager, fullgraph=False)(fn)(x)
+        self.assertEqual(expected, out)
+        self.assertEqual(len(eager.graphs), 0)
+
+    @unittest.expectedFailure
+    def test_disable___enter__(self):
```

Comment:
I thought about this some more. We should be able to support this (via graph breaks) easily. If I am understanding this correctly, the problem is  that the `__enter__` is run in cpython and the `__exit__` is run in Dynamo.

We should just graph break if `__enter__` or `__exit__` is directly called on a GenericContextManager that is backed by contextlib.contextmanager.

Reply:
> If I am understanding this correctly, the problem is that the __enter__ is run in cpython and the __exit__ is run in Dynamo.

No. The problem is that Dynamo still traces `whoo` and `h`. It does because upon re-entering the function, it loses the information that `g` is a `FunctionDecoratedByContextlibContextManagerVariable`.

Reply:
Would always graph breaking on `ctx.__enter__/ctx.__exit__` calls fix this test case?

Reply:
Oh this is this thing where we want recursive skip

Reply:
Updated proposal: graph break on next(_GenerateContextManager.gen)

https://github.com/python/cpython/blob/5dd775bed086909722ec7014a7c4f77a35f74a80/Lib/contextlib.py#L141

This seems plausible, dynamo could turn the _GeneratorContextManager into a _GeneratorContextManagerVariable or something

Reply:
> Would always graph breaking on ctx.__enter__/ctx.__exit__ calls fix this test case?

Maybe. But code with fullgraph=True wouldn't compile if we go with this route. I think I got it working, but we should discuss it more thoroughly during the next meeting.